### DISCUSSION TITLE
RTU Server example : packet length error

### DISCUSSION THREAD

**Discussion by manu-t** [eModbus/eModbus#189] [Q&A] [answered]
Hi,
I am trying the Modbus RTU server example FC03 on ESP32 devkit board. The example seems to work fine but I have packet length errors every time I send a packet:
Packet send from a ftdi cable to serial2: `01 03 00 02 00 01 25 CA` (get 2nd value)
received from RTU server: `01 03 02 00 02 39 85`
(which is expected)
but on the serial console I have this message:
`ModbusServerRTU.cpp  [ 198] serve: RTU receive: E5 - Packet length error`
Do you think that the fact that I'm using directly UART 2 with a FTDI cable (and no RS485 transceiver) could do that?
I'm just discovering the library I don't know exactly how to send debug traces that I've seen on other discussion, but I will post them when I know how to do it.

**Comment by manu-t** [comment]
The code that I'm using (I inverted the UART2 pins):
```
// =================================================================================================
// eModbus: Copyright 2020 by Michael Harwerth, Bert Melis and the contributors to ModbusClient
//               MIT license - see license.md for details
// =================================================================================================
// Includes: <Arduino.h> for Serial etc., WiFi.h for WiFi support
#include <Arduino.h>
#include "HardwareSerial.h"

// Modbus server include
#include "ModbusServerRTU.h"

#define LOCAL_LOG_LEVEL LOG_LEVEL_VERBOSE
#include "Logging.h"

// Create a ModbusRTU server instance listening on Serial2 with 2000ms timeout
ModbusServerRTU MBserver(Serial2, 2000);

// FC03: worker do serve Modbus function code 0x03 (READ_HOLD_REGISTER)
ModbusMessage FC03(ModbusMessage request) {
  uint16_t address;           // requested register address
  uint16_t words;             // requested number of registers
  ModbusMessage response;     // response message to be sent back

  // get request values
  request.get(2, address);
  request.get(4, words);

  // Address and words valid? We assume 10 registers here for demo
  if (address && words && (address + words) <= 10) {
    // Looks okay. Set up message with serverID, FC and length of data
    response.add(request.getServerID(), request.getFunctionCode(), (uint8_t)(words * 2));
    // Fill response with requested data
    for (uint16_t i = address; i < address + words; ++i) {
      response.add(i);
    }
    Serial.println("reponse OK");
  } else {
    // No, either address or words are outside the limits. Set up error response.
    response.setError(request.getServerID(), request.getFunctionCode(), ILLEGAL_DATA_ADDRESS);
  }
  return response;
}

// Setup() - initialization happens here
void setup() {
// Init Serial monitor
  Serial.begin(115200);
  while (!Serial) {}
  Serial.println("__ OK __");

// Init Serial2 connected to the RTU Modbus
// (Fill in your data here!)
  Serial2.begin(19200, SERIAL_8N1, GPIO_NUM_16, GPIO_NUM_17);

// Register served function code worker for server 1, FC 0x03
  MBserver.registerWorker(0x01, READ_HOLD_REGISTER, &FC03);

// Start ModbusRTU background task
  MBserver.start();
}

// loop() - nothing done here today!
void loop() {
  delay(10000);
}
```

**Comment by Miq1** [comment]
It should not matter you have no RS485 adapter in between - as seen from the ESP32, all is serial comms only (with the exception of the DE/RE line, in case).
Are you using the Arduino IDE or PlatformIO for development? On PlatformIO it is easier to set the logging flags: just add a ``-DLOG_LEVEL=6`` to the ``build_flags`` in ``platformio.ini``. On the Arduino IDE you might have to modify the ``Logging.h`` file directly to set the ``LOG_LEVEL`` to the desired default state. I heard it would be possible to set compiler flags in the IDE as well, but never tried to find that myself.
Logging output will be on the Serial Monitor.

**Reply by manu-t** [reply]
Thanks for your help!
I'm using PlatformIO (which I'm discovering along with the library). It seems that I constantly receive a byte:
```
[D] Received packet: @3FFB998C/1:
  | 0000: E0                                                |.               |
[D] Received packet: @3FFB99A0/1:
  | 0000: E0                                                |.               |
[D] Received packet: @3FFB998C/1:
  | 0000: E0   
```
I probably have to put an oscilloscope probe on this...

When I send a correct packet, the debug is:
```
[D] Raw buffer received: @3FFB38C8/8:
  | 0000: 01 03 00 02 00 01 25 CA                           |......%.        |
[D] Received packet: @3FFB99E0/6:
  | 0000: 01 03 00 02 00 01                                 |......          |
[D] 198214| ModbusServerRTU.cpp  [ 141] serve: Request received.
[D] 198225| ModbusServer.cpp     [  30] getWorker: Worker found for 01/03
[D] 198225| ModbusServerRTU.cpp  [ 146] serve: Callback found.
reponse OK
[D] 198235| ModbusServerRTU.cpp  [ 154] serve: Callback called.
[D] Sent packet: @3FFB99F8/5:
  | 0000: 01 03 02 00 02                                    |.....           |
[D] 198249| ModbusServerRTU.cpp  [ 190] serve: Response sent.
[D] Raw buffer received: @3FFB38C8/1:
  | 0000: 00                                                |.               |
[D] Received packet: @3FFB99F8/1:
  | 0000: E5                                                |.               |
␛[1;33m[E] 198282| ModbusServerRTU.cpp  [ 198] serve: RTU receive: E5 - Packet length error
␛[0m[D] Received packet: @3FFB9A0C/1:
  | 0000: E0                                                |.               |
```
which is fine, then I get the packet length error with a null byte

**Reply by Miq1** [reply]
Yeah, whatever you are using as a client is sending a single 00 byte, so it seems. The server cannot make heads or tails from it and will log an E5 error.
Further on there are E0 bytes logged. This is done after a complete receive cycle without any data having been received and can safely be (and in fact _is_) ignored for a server.

You will need to find out why these 00 bytes are sent by the client. These can be sent explicitly, or you got some unstable UART lines with noise on DTR set or similar.

**Reply by manu-t** [reply]
> You will need to find out why these 00 bytes are sent by the client. These can be sent explicitly, or you got some unstable UART lines with noise on DTR set or similar.


It seems some kind of biasing is mandatory on the Rx side as I still receive fake bytes no matter what I do (I've shut down the X-ctu serial terminal, disconnected the ftdi header, and still receive phantom bytes) => I expected the Tx line on the ftdi adapter would keep stable logical states for the Rx ESP32 side, but it's not the case

**Reply by Miq1** [reply]
I never experienced that before, but "live and learn", as they say.

**Comment by manu-t** [comment]
It gets stranger and stranger: I've try different ESP32 boards (dev kit, Huzzah32) and different UART configs (direct FTDI cable, RS485 driver, ...) but still random byte at Rx. Pullup/pulldown resistor (real resistor and internal weak PU in the ESP32) on the Rx line does nothing.
If someone else has this problem, please let me know...

**Reply by Miq1** [reply]
You may need to do proper biasing on the RS485 side, not the UART. It is advised to have pullup A+/pulldown B- at *one side of the RS485 bus* and 120 Ohms termination at each bus device between A+ and B-. The latter is mostly integrated already in the RS485 adapter modules.

**Reply by bertmelis** [reply]
Do you mean you get these random bytes when you directly connect via UART, so no converter or anything in between? That is really curious and something that I'd like to see on a scope or logic analyzer.

**Reply by manu-t** [reply]
I just connected a FTDI USB cable on the Rx and Tx pin (and GND). Here is the logic analyzer capture:
![image](https://user-images.githubusercontent.com/17081912/168585912-d702f79d-5b56-477c-b5f2-386ace821d1a.png)
The Rx line seems pretty clean to me.
Here is a zoom out where you can see the response about 55ms after:
![image](https://user-images.githubusercontent.com/17081912/168586078-6dadf899-1cbe-483d-b94d-99a3e5d6c246.png)

I will try with an oscilloscope.

**Reply by manu-t** [reply]
On the scope, it pretty clean too:
![image](https://user-images.githubusercontent.com/17081912/168596614-9123b69a-eb52-4d04-8e48-9d8f3ea4dfad.png)
![image](https://user-images.githubusercontent.com/17081912/168596651-8ca3ad42-c23d-45b8-8584-9410d194b8cf.png)
![image](https://user-images.githubusercontent.com/17081912/168596710-9a697db2-c5c3-4040-8514-6e81f3312363.png)
![image](https://user-images.githubusercontent.com/17081912/168598685-7e736107-985f-4fb9-be0f-fa4daafac869.png)

**Comment by Miq1** [comment]
Looks good indeed. 

That sounds a bit like that we are currently investigating on the [issue #186](https://github.com/eModbus/eModbus/issues/186), if in a different flavour. Something is throwing in bytes on the core level, so it seems.

**Reply by manu-t** [reply]
let me know if I can help one way or the other...

**Reply by Miq1** [reply]
Problem is, like you did the serial data looks okay on the line, whereas that we get through ``serial.read()`` is not - sometimes. That means that the data is changed in the core, on the way from the FIFO to the UART buffer. That is nothing we can do much about. If only we had a test case that will provoke the error every single time it is run...

**Reply by manu-t** [reply]
well, it's consistent with my setup: every timeout (2000ms), I get a packet length error

**Reply by Miq1** [reply]
That is different. Can you provide us with a fresh verbose dump of that happening?

**Reply by manu-t** [reply]
Here is the latest log:
```
__ OK __
[D] 28| ModbusServer.cpp     [  16] registerWorker: Registered worker for 01/03
[D] 29| RTUutils.cpp         [ 146] UARTinit: Serial2 FIFO threshold set to 1 (was 120)
[D] 39| ModbusServerRTU.cpp  [  84] start: Server task 1073453068 started. Interval=1750
[D] Received packet: @3FFB9974/1:
  | 0000: E0                                                |.               |
[D] Received packet: @3FFB9988/1:
  | 0000: E0                                                |.               |
[D] Received packet: @3FFB9974/1:
  | 0000: E0                                                |.               |
[D] Received packet: @3FFB9988/1:
  | 0000: E0                                                |.               |
[D] Received packet: @3FFB9974/1:
  | 0000: E0                                                |.               |
[D] Received packet: @3FFB9988/1:
  | 0000: E0                                                |.               |
[D] Received packet: @3FFB9974/1:
  | 0000: E0                                                |.               |
[D] Received packet: @3FFB9988/1:
  | 0000: E0                                                |.               |
[D] Received packet: @3FFB9974/1:
  | 0000: E0                                                |.               |
[D] Received packet: @3FFB9988/1:
  | 0000: E0                                                |.               |
[D] Received packet: @3FFB9974/1:
  | 0000: E0                                                |.               |
[D] Received packet: @3FFB9988/1:
  | 0000: E0                                                |.               |
[D] Received packet: @3FFB9974/1:
  | 0000: E0                                                |.               |
[D] Received packet: @3FFB9988/1:
  | 0000: E0                                                |.               |
[D] Received packet: @3FFB9974/1:
  | 0000: E0                                                |.               |
[D] Received packet: @3FFB9988/1:
  | 0000: E0                                                |.               |
[D] Received packet: @3FFB9974/1:
  | 0000: E0                                                |.               |
[D] Received packet: @3FFB9988/1:
  | 0000: E0                                                |.               |
[D] Received packet: @3FFB9974/1:
  | 0000: E0                                                |.               |
[D] Received packet: @3FFB9988/1:
  | 0000: E0                                                |.               |
[D] Received packet: @3FFB9974/1:
  | 0000: E0                                                |.               |
[D] Received packet: @3FFB9988/1:
  | 0000: E0                                                |.               |
[D] Received packet: @3FFB9974/1:
  | 0000: E0                                                |.               |
[D] Received packet: @3FFB9988/1:
  | 0000: E0                                                |.               |
[D] Received packet: @3FFB9974/1:
  | 0000: E0                                                |.               |
[D] Received packet: @3FFB9988/1:
  | 0000: E0                                                |.               |
[D] Received packet: @3FFB9974/1:
  | 0000: E0                                                |.               |
```

**Reply by manu-t** [reply]
and when I send a packet from the Xctu terminal:
```
[D] Raw buffer received: @3FFB38C8/8:
  | 0000: 01 03 00 02 00 01 25 CA                           |......%.        |
[D] Received packet: @3FFB999C/6:
  | 0000: 01 03 00 02 00 01                                 |......          |
[D] 667471| ModbusServerRTU.cpp  [ 141] serve: Request received.
[D] 667482| ModbusServer.cpp     [  30] getWorker: Worker found for 01/03
[D] 667482| ModbusServerRTU.cpp  [ 146] serve: Callback found.
[D] 667493| ModbusServerRTU.cpp  [ 154] serve: Callback called.
[V] Callback response: @3FFB99C8/5:
  | 0000: 01 03 02 00 02                                    |.....           |
[D] Sent packet: @3FFB99E0/5:
  | 0000: 01 03 02 00 02                                    |.....           |
[D] 667514| ModbusServerRTU.cpp  [ 190] serve: Response sent.
[D] Raw buffer received: @3FFB38C8/1:
  | 0000: 00                                                |.               |
[D] Received packet: @3FFB99B4/1:
  | 0000: E5                                                |.               |
␛[1;33m[E] 667547| ModbusServerRTU.cpp  [ 198] serve: RTU receive: E5 - Packet length error
␛[0m[D] Received packet: @3FFB998C/1:
  | 0000: E0                                                |.               |
[D] Received packet: @3FFB99A0/1:
  | 0000: E0                                                |.               |
[D] Received packet: @3FFB998C/1:
```

**Comment by Miq1** [comment]
Yeah, that is again a stray 0x00 byte, this time trailing behind the message. Again I would think that is a ghost byte generated by the DE/RE toggle - but you wrote you have no RS485 adapter attached at all?

**Reply by manu-t** [reply]
Just the ftdi cable, however, I specified GPIO 23 as DE/RE for when I use a RS485 adapter :
```
ModbusServerRTU MBserver(Serial2, 2000, 23);
```

**Comment by Miq1** [comment]
But that will have no impact on the communication as long as no real adapter is used. 

I am stunned...

The byte is received ony a few milliseconds after the response has been sent, and it obviously is tied to that response (else more of these 0x00 bytes would pop up in idle state). Read and write FIFOs are separate for the UARTs, so it is unlikely we are writing something into the receive FIFO while sending.

**Comment by manu-t** [comment, marked as answer]
Hello,
There is no more packet length error with the workaround from the v1.5.0
Thank you for the good work! 
I added 
```
MBserver.skipLeading0x00();
```
at the end of the `setup()` function

**Reply by Miq1** [reply]
Weeell... Although I am glad to read that this could cure your problem, it still is an ugly hack to remove a symptom from a problem lying below. 😉

**Reply by manu-t** [reply]
pile of dirty hacks, as all good project should be  :laughing:

**Reply by bertmelis** [reply]
😬

**Reply by manu-t** [reply]
Just in case (don't get me wrong): I was referring to my project and not your library ;-)
